### ISSUE TITLE
Generating an x-ray on a line chart on big tables will generate a massive x-ray that will lock up the browser

### ISSUE THREAD

**Issue by paoliniluis** [metabase/metabase#43258]
### Describe the bug

A customer has a very big table, which they use to generate a simple line chart. They found that when they use an x-ray on a data point the x-ray generated is massive

### To Reproduce

1) set up Metabase with a postgres db
2) load this massive dataset https://drive.google.com/file/d/1rbp2or3Kdoc6Fy-9j1m2HCA4NvOEa2A9/view?usp=sharing
3) do a very simple question like
![image](https://github.com/metabase/metabase/assets/1711649/8598b4d9-52c5-4648-a232-b83d76485214)

4) then click on any data point ->automatic insights- >x-rays
![image](https://github.com/metabase/metabase/assets/1711649/d844b078-af1d-4d3f-b5ca-2581fbff53f9)

you'll get an ultra massive dashboard that locks up the browser and probably the server as well

### Expected behavior

we need product on this

### Logs

NA

### Information about your Metabase installation

```JSON
since the very beginning
```

### Severity

P1

### Additional context

NA

**Comment by darksciencebase** [comment]
[thread mentioning slowness in the tests as well](https://metaboat.slack.com/archives/C0641E4PB9B/p1718637304256719). if this turns out to be a separate issue, we should create a new ticket

**Comment by piranha** [comment]
My findings: the `automagic-dashboards/adhoc/.../cell/...` API returns a dashboard with 311 cards (16 kb json) in ~5 seconds, and then `automagic-dashboards/adhoc/.../query_metadata` returns a table with 500 fields (23 kb json) in ~10 seconds. 

Browser is perfectly responsive during this time, but then locks up to generate queries for every card at the same time. If I profile JS, it indicates that LOTS of functions is called (and the functions are not small either, each one of them takes about 20-23 ms).

<img width="1512" alt="image" src="https://github.com/metabase/metabase/assets/6553/69ef7cf4-ca33-463b-adbe-e5261b8a2074">

For example, DashCardInner, after following all of nesting, ends up calling `getUrlWithParameters` (takes 23 ms), which in turns calls `getUrl` (which spends 11.1ms in `fromLegacyQuery`) and `query` (which also spends 11.1ms in `fromLegacyQuery`). Most of the time of `fromLegacyQuery` is spent in `metabase$lib$schema$join_ref_error_for_stages`.

This issue can be tackled from three sides:

- Limit amount of work to be done, i.e. don't render so many cards at once (or least do that with pagination)
  - Either from server side, this can also speed up initial responses, and will help not to lock UI as much
  - Or for client side, just render whatever the client will see first and then reveal/fetch data on scroll, that will also help with UI locking and the API will stay simple with no pagination as it is right now
- Optimize `metabase$lib$schema$join_ref_error_for_stages` (and possibly other functions), it needs to be much faster to be used like it is used right now.

Just in case, [profile is here](https://drive.google.com/file/d/1-ekg0bH17TmSb9Hecj9W33kUgtzI9y0w/view?usp=sharing), if you want to see it without reproducing (profiler misbehaves seeing this page too 🤣).

**Comment by kamilmielnik** [comment]
> For example, DashCardInner, after following all the nesting, ends up calling `getUrlWithParameters` (takes 23 ms),

This was due to [Copy question link from dashboard](https://www.notion.so/metabase/Copy-question-link-from-dashboard-702fe6ec9c874f12bb2cda19dbbf6f47) feature (PR: #43816).

We've added optimization to compute chart title link only on mouseenter and focus: #45095

I'm not closing the issue yet, as 300-card dashboards are still slow. The app does not freeze anymore though. See #45095 for videos.
See [Slack](https://metaboat.slack.com/archives/C05NXACAG1G/p1719922327712559) for details.
